File: java/json-transform/src/test/java/co/nlighten/jsontransform/functions/TransformerFunctionConcatTest.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.BaseTest;
import org.junit.jupiter.api.Test;

public class TransformerFunctionConcatTest extends BaseTest {
    @Test
    void testObjectFunctionConcat() {
        var arr = new String[] { "a", "b", "c"};
        var arr2 = new String[] { "d", "e", "f"};
        var arr3 = new String[][] { arr, arr2};
        assertTransformation(arr3, fromJson("""
{
  "$$concat": [ "$[0]", "$[1]" ]
}
"""), fromJson("""
["a","b","c","d","e","f"]"""));

        // keep nulls
        assertTransformation(arr3, fromJson("""
{
  "$$concat": [ ["a",null,"c"], ["d","e",null] ]
}
"""), fromJson("""
["a",null,"c","d","e",null]"""));

        // skip nulls
        assertTransformation(arr, fromJson("""
{
  "$$concat": [ ["a","b","c"], null ]
}
"""), fromJson("""
["a","b","c"]"""));

        // append non nulls
        assertTransformation(arr, fromJson("""
{
  "$$concat": [ ["a","b","c"], "d", ["e"] ]
}
"""), fromJson("""
["a","b","c","d","e"]"""));

        // blind concat
        assertTransformation(fromJson("[ [\"a\",\"b\",\"c\"], \"d\", [\"e\"] ]"), fromJson("""
{
  "$$concat": "$"
}
"""), fromJson("""
["a","b","c","d","e"]"""));
    }
}

File: javascript/json-transform/src/__tests__/functions/TransformerFunctionConcat.test.ts
import { describe, test } from "vitest";
import { assertTransformation } from "../BaseTransformationTest";

describe("TransformerFunctionConcat", () => {
  test("object", async () => {
    const arr = ["a", "b", "c"];
    const arr2 = ["d", "e", "f"];
    const arr3 = [arr, arr2];
    await assertTransformation(arr3, { $$concat: ["$[0]", "$[1]"] }, ["a", "b", "c", "d", "e", "f"]);

    // keep nulls
    await assertTransformation(
      arr3,
      {
        $$concat: [
          ["a", null, "c"],
          ["d", "e", null],
        ],
      },
      ["a", null, "c", "d", "e", null],
    );

    // skip nulls
    await assertTransformation(arr, { $$concat: [["a", "b", "c"], null] }, ["a", "b", "c"]);

    // append non nulls
    await assertTransformation(arr, { $$concat: [["a", "b", "c"], "d", ["e"]] }, ["a", "b", "c", "d", "e"]);

    // blind concat
    await assertTransformation([["a", "b", "c"], "d", ["e"]], { $$concat: "$" }, ["a", "b", "c", "d", "e"]);
  });
});
